### PR TITLE
test/e2e: allow release PRs to pass CI

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
* test/e2e/main_test.go: use testArgs struct for test args; add local-repo flag to specify the local repo path if the SDK is not in $GOPATH/src/github.com/operator-framework/operator-sdk

* test/e2e/memcached_test.go: remove any 'replace' directives for the SDK repo in memcached-operator's go.mod file before adding one for the e2e test

* hack/tests/e2e-{ansible,helm}.sh: same as above.

**Motivation for the change:** release PR's will [fail](https://github.com/operator-framework/operator-sdk/pull/1445) because the release tag they are pinning in `go.mod` does not exist yet. This change removes the release tag pin (ex. `replace github.com/operator-framework/operator-sdk => github.com/operator-framework/operator-sdk vX.Y.Z`) before adding a `replace` line for the PR sha/branch. This is essentially what the e2e code was doing when `dep` was the default dependency manager.

The command line `-local-repo` is necessary for devs working on the SDK repo in a path other than `$GOPATH/src/replace github.com/operator-framework/operator-sdk`. The `$GOPATH` requirement still remains, but will be removed soon, at which point `-local-repo` will be especially useful.

### CODE REVIEW

**Comment by joelanford** [comment]
I think we may also need to account for the Helm and Ansible e2e tests. They also failed in the `v0.8.1` release PR. It looks like its the `operator-sdk migrate` tests that fail due to the missing tag.

See https://travis-ci.org/operator-framework/operator-sdk/builds/535849739

**Comment by estroz** [comment]
@joelanford added a `go mod edit -dropreplace` before vendoring/adding a replace for ansible and helm e2e.

**Comment by estroz** [comment]
/cc @joelanford @AlexNPavel @LiliC @hasbro17 

PTAL

**Comment by estroz** [comment]
On further consideration, I've realized this PR depends on #1470 and removing the SDK's dependency on `vendor/`.

`operator-sdk new` will error in Go e2e tests in a release PR because the release tag does not exist; we currently have no way of preventing the SDK from calling `go mod vendor` so this will occur unless we skip vendoring on project initialization. I propose we wait until #1470 is merged, and make a separate PR that adds the [`--vendor`](https://github.com/operator-framework/operator-sdk/pull/1475/files#diff-e5b830b542b50cf7d9d7d8c43a1f792fR64) flag from #1475 so we can modify the `go.mod` file prior to running downstream e2e setup, like we do in `hack/tests/e2e-{ansible,helm}.sh`.

@hasbro17 @AlexNPavel @joelanford @LiliC @theishshah wdyt?

**Comment by LiliC** [comment]
It seems like the two PRs in question are merged, is this ready for review again?

**Comment by estroz** [comment]
@LiliC there are a few updates I need to make first.

File: hack/tests/e2e-ansible.sh [hunk 1]
```diff
@@ -133,6 +133,14 @@ then
     exit 1
 fi
 
+# Remove any "replace" and "require" lines for the SDK repo before vendoring
+# in case this is a release PR and the tag doesn't exist yet. This must be
+# done without using "go mod edit", which first parses go.mod and will error
+# if it doesn't find a tag/version/package.
+# TODO: remove SDK repo references if PR/branch is not from the main SDK repo.
+SDK_REPO="github.com/operator-framework/operator-sdk"
+sed -E -i 's|^.*'"$SDK_REPO"'.*$||g' go.mod
```
Comment (joelanford) [inline hack/tests/e2e-ansible.sh]: You might be able to use `go mod edit -droprequire` and `go mod edit -dropreplace` instead. Those flags don't seem to cause tag/version/package lookups.

File: test/e2e/memcached_test.go [hunk 2]
```diff
@@ -125,7 +130,22 @@ func TestMemcached(t *testing.T) {
...
 			}()
 		}
-		writeGoModReplace(t, sdkRepo, replace.repo, replace.ref)
+		modBytes, err := ioutil.ReadFile("go.mod")
+		if err != nil {
+			t.Fatalf("Failed to read go.mod: %v", err)
+		}
+		// Remove SDK repo dependency lines so we can parse the modfile.
+		sdkRe := regexp.MustCompile(`.*github\.com/operator-framework/operator-sdk.*`)
+		modBytes = sdkRe.ReplaceAll(modBytes, nil)
```
Comment (joelanford) [inline test/e2e/memcached_test.go]: You might be able to use `modFile.DropReplace` and `modFile.DropRequire` here, similar to my comment using the equivalent CLI flags in the ansible and helm scripts.

File: hack/tests/e2e-ansible.sh (hunk 1 above)
Comment (estroz) [reply hack/tests/e2e-ansible.sh]: I'm fairly certain any `go mod edit` command will perform lookups:
```
$ GO111MODULE=on go mod edit -droprequire github.com/operator-framework/operator-sdk
go: errors parsing go.mod:
/home/estroz/go/src/github.com/test-org/metrics-operator/go.mod:22: invalid module version "master": version must be of the form v1.2.3
```

File: test/e2e/memcached_test.go (hunk 2 above)
Comment (estroz) [reply test/e2e/memcached_test.go]: Same as above.

File: hack/tests/e2e-ansible.sh (hunk 1 above)
Comment (joelanford) [reply hack/tests/e2e-ansible.sh]: Ahh. I didn't test with branch names (just non-existent tag versions). Oh well.

**Comment by hasbro17** [comment]
This is more for us but do we need to mention the `--local-repo` flag in the [`doc/dev/running-the-testing.md`](https://github.com/operator-framework/operator-sdk/blob/master/doc/dev/testing/running-the-tests.md) file?

I know it's probably not glanced at often enough but we can add a line on how to pass the `--local-repo` flag to the makefile args.
https://github.com/operator-framework/operator-sdk/blob/master/doc/dev/testing/running-the-tests.md#running-the-tests

**Pull request merged**